src/programmatic_policy_learning/approaches/residual_approach.py

```diff
@@ -124,62 +132,142 @@ def load(self, path: str) -> None:
+            return np.asarray(a, dtype=np.float32)
+
+        self._base_fn: Callable[[np.ndarray], np.ndarray] = _base_fn
+
+        train_env = self._env_factory(train_env_instance)
+        train_env.reset(seed=seed)
+        self._train_env = ResidualActionWrapper(train_env, self._base_fn)
         self._backend = _SB3Backend(
-            backend,
-            env=self._env,
+            name=cast(_BackendName, backend),
+            env=self._train_env,
             seed=seed,
             lr=lr,
             noise_std=noise_std,
             verbose=verbose,
         )
-        self._total = total_timesteps
+        self._total = int(total_timesteps)
 
-        # Keep a Box-typed handle to action space for mypy (low/high/dtype).
-        if not isinstance(self._action_space, Box):
-            raise TypeError("ResidualApproach requires a Box action space.")
-        self._act_box: Box = cast(Box, self._action_space)
+        self._is_trained: bool = False
+        self._train_before_eval = bool(train_before_eval)
+        self._last_obs: Optional[np.ndarray] = None
 
+    # ------- pipeline hooks used by the Hydra runner -------
     def train(self) -> None:
-        """Train the residual policy."""
         self._backend.learn(total_timesteps=self._total)
+        self._is_trained = True
+
+    def reset(self, obs: np.ndarray, info: dict) -> None:
+        if self._train_before_eval and not self._is_trained:
+            self.train()
+        self._last_observation = np.asarray(obs, dtype=np.float32)
+        self._last_obs = self._last_observation
 
-    def _get_action(self) -> _ActType:
+    def _get_action(self) -> np.ndarray:
         """Return clipped action = base + residual."""
-        # pylint: disable=protected-access
+        assert self._last_observation is not None, "Call reset() before _get_action()."
         obs = np.asarray(self._last_observation, dtype=np.float32)
-        base = self._base_fn(obs)  # fixed/base policy
-        residual = self._backend.predict(obs)  # learned residual
+        base = self._base_fn(obs)
+        residual = self._backend.predict(obs)
+        total = np.clip(base + residual, self._act_box.low, self._act_box.high)
+        return total.astype(self._act_box.dtype)
+
+    def step(self) -> np.ndarray:
+        assert self._last_obs is not None, "Call reset() before step()."
+        obs = np.asarray(self._last_obs, dtype=np.float32)
+        base = self._base_fn(obs)
+        residual = self._backend.predict(obs)
         total = np.clip(base + residual, self._act_box.low, self._act_box.high)
-        return total.astype(self._act_box.dtype)  # type: ignore[return-value]
+        return total.astype(self._act_box.dtype)
+
+    def update(self, obs: np.ndarray, reward: float, done: bool, info: dict) -> None:
+        self._last_observation = np.asarray(obs, dtype=np.float32)
+        self._last_obs = self._last_observation
 
+    def test_policy_on_envs(
```

Comment:
in my mind, testing is something that should be done "to" the policy, rather than something that the policy "does itself". can we move this method outside the class?

Reply:
Oh I see! Sorry, I am realizing things about the existing code that I missed before. Let's leave this alone for now but maybe revisit later